### PR TITLE
gh-104050: Argument Clinic: Annotate output_templates()

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-104050

### CODE REVIEW

**Comment by AlexWaygood** [comment]
The mypy error indicates we might have the wrong annotation for `Parameter.converter`

**Comment by erlend-aasland** [comment]
Thanks for 6891e14096f6fc91424c9070da945aec9f8a0b89! The last error puzzles me, though, and I have no bandwidth for this today; if you find a solution, feel free to land this :)

**Comment by AlexWaygood** [comment]
> The last error puzzles me, though, and I have no bandwidth for this today; if you find a solution, feel free to land this

Yeah, I can't immediately tell what the issue is here. I'm investigating :)

Mypy doesn't have a _great_ understanding of multiple inheritance, so this may be a false positive. But not sure.

**Comment by erlend-aasland** [comment]
Aaah, perhaps it is line 801 and 802:

```python
assert isinstance(parameters[0].converter, self_converter)
del parameters[0]
```

My guess: mypy uses only the information provided in line 801, but is not smart enough to see that line 802 invalidates that assert, so it continues to think `parameters[0]` is asserted as a `self_converter` instance.

**Comment by AlexWaygood** [comment]
> Aaah, perhaps it is line 801 and 802:
> 
> ```python
> assert isinstance(parameters[0].converter, self_converter)
> del parameters[0]
> ```
> 
> My guess: mypy uses only the information provided in line 801, but is not smart enough to see that line 802 invalidates that assert, so it continues to think `parameters[0]` is asserted as a `self_converter` instance.

Bingo! I'll push a fix.

**Comment by AlexWaygood** [comment]
f9cb537 look good to you?

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -792,11 +792,14 @@ def docstring_for_c_string(
...
+    def output_templates(
+            self,
+            f: Function
+    ) -> dict[str, str]:
         parameters = list(f.parameters.values())
         assert parameters
-        assert isinstance(parameters[0].converter, self_converter)
-        del parameters[0]
+        first_param = parameters.pop(0)
+        assert isinstance(first_param.converter, self_converter)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:802]: We could possibly code-golf this down a bit more, if we wanted to:

```suggestion
        assert isinstance(parameters.pop(0).converter, self_converter)
```

Comment (erlend-aasland) [reply Tools/clinic/clinic.py:802]: IMO, doing too many things in one line makes the code harder to read; I prefer the explicitness of the existing variant :)

**Comment by erlend-aasland** [comment]
> Does [f9cb537](https://github.com/python/cpython/commit/f9cb5371a794ac269623f025d4b6a703f2993c26) look good to you?

Neat!

**Review by AlexWaygood — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks!

**Pull request merged**